design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,112 @@
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
+
+- Complex to collect all existing metrics and alerts for automatic documentation
+
+- High levels of intertwined monitoring and business logic code
+
+### Goals
+
+- Decouple monitoring logic from business logic
+- Have a centralized place for the configuration of monitoring resources
+- Encapsulate the monitoring best-practices and the common patterns into a
+library and have it as a dependency for all KubeVirt components
+- Have a declarative way to define metrics and alerts
+- Keep monitoring code and utilities easy to maintain and evolve
+- Document how to use the monitoring library for adding new metrics and rules
+- Have a structure and tools to accurately and easily generate monitoring
+documentation, lint metrics and alerts, define allow-list, deny-list and
+opt-in metrics and other future features without having to change the code in
+multiple places
+
+### Non-Goals
+
+- Update any metrics or alerts
+- Apply these ideas to events in this phase
+- Create a new component to run in the cluster alongside other components to
+collect metrics and create alerts
+
+## How
+
+### General relationship diagram:
+
+<img src="relationship.png" alt="relationship diagram" width="800"/>
+
+With this refactor, we intend to move out monitoring logic from all the KubeVirt
+components' business logic code. As we see in the above diagram, each KubeVirt
+component will now import a package in the `kubevirt/monitoring` repository.
+Each package will contain, for the corresponding component, the declaration and
+registration of metrics, the generation of the `PrometheusRule` resource with
+the alerts and recording rules, and the automatically generated documentation
+information.
+
+In turn, each component package in `kubevirt/monitoring`, will use a generic
+package for operator observability, abstracting all common operations between
+the components (such as registering them in `controller-runtime` metrics
+`Registry`), deduplicating code, and providing a way to define the metrics,
+recording rules , and alerts in a declarative manner.
+
+### Relationship diagram for each component monitoring code:
+
+<img src="component_monitoring_relationship.png" alt="relationship diagram" width="800"/>
+
+Each component package in `kubevirt/monitoring`, will have a `metrics/` package
+with the metrics declarations, the methods to update/consult the metrics values
+and the collectors logic. The `rules/` package will also have the declarative
+configuration of the resources, in this case, alerts and recording rules.
```

Comment:
Can you show an example of this declarative configuration? I think it would be much easier to picture this solution that way

Reply:
added

Reply:
Thanks! Much clearer now!